builder/main.py

```
@@ -133,6 +150,264 @@ def fetch_fs_size(env):
+        print(f"Error building SPIFFS image: {e}")
+        return 1
+
+
+def build_fatfs_image(target, source, env):
+    """Build FatFS filesystem image with ESP32 Wear Leveling support."""
+    source_dir = str(source[0])
+    target_file = str(target[0])
+    fs_size = env["FS_SIZE"]
+    sector_size = env.get("FS_SECTOR", 4096)
+
+    fat_fs_size = wl_info['fat_size']
```

Ruff (0.14.10)

Undefined name `wl_info`

(F821)